Fix part of #6271: Refactor solution update method to accept domain object instead of dict.

FlorinBalint opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
#6271 wants to remove the usage of dicts from the updates of "domain" objects.
This part addresses the update_interaction_solution method, from the Solution domain class.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PR CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

FlorinBalint committed:
Fix part of #6271: Refactor solution update method to accept domain
object instead of dict.

FlorinBalint committed:
Fix part of #6271: Refactor solution update method to accept domain
object instead of dict.

FlorinBalint committed:
Fix part of #6271: Refactor solution update method to accept domain
object instead of dict.

seanlip commented on core/domain/exp_domain_test.py:
```diff
@@ -717,14 +717,18 @@ def test_validation(self):
-        })
+            }
+        }
+        solution = state_domain.Solution.from_dict(
```
Perhaps here you could construct the Solution object directly, using the constructor. Ditto elsewhere. In general we should try to avoid using dicts except at endpoints of the app.

FlorinBalint replied:
Done.
The service falls under the "endpoint case" (naming sort of states that) right?

seanlip replied:
No, I don't think so -- "endpoint" here means the controller layer, i.e. the part in core.controllers that is just before self.render_json(), since that is the point at which we need to serialize stuff before sending it to the frontend. (Sorry for not being clear before.)

seanlip commented on core/domain/state_domain.py:
```diff
@@ -1796,29 +1796,20 @@ def update_interaction_hints(self, hints_list):
-
-        Raises:
-            Exception: 'solution_dict' is not a dict.
+            solution: object or None. The Solution object.
```
The typeinfo should be Solution, not object. (The latter is too generic.)

seanlip commented on core/domain/state_domain_test.py:
```diff
@@ -684,14 +688,17 @@ def test_hints_validation(self):
         init_state.update_interaction_hints(hints_list)
 
-        solution = {
+        solution_dict = {
```
Here (and elsewhere in this file) also, don't use dicts, construct the solution object directly.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -2152,13 +2152,16 @@ def _create_valid_question_data(self, default_dest_state_name):
                 'html': '<p>This is a solution.</p>'
             }
         }
+        solution = state_domain.Solution.from_dict(
```
Ditto.

seanlip reviewed (changes requested):
Thanks @FlorinBalint -- looks pretty good! Just one comment, repeated throughout (+ a note about typeinfo).

FlorinBalint committed:
Merge branch 'interaction_update_6271' of https://github.com/FlorinBalint/oppia into interaction_update_6271

FlorinBalint committed:
Use the constructor directly instead of the 'from_dict' method.

FlorinBalint committed:
Merge remote-tracking branch 'upstream/develop' into develop

FlorinBalint committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

FlorinBalint commented:
I forgot to update one usage in state_domain_test. Also merged with upstream in the same commit.

After looking into what happens with the [0, 0] answer type (it gets normalized):
Should we update the from_dict to skip the normalization ? 
(since the constructor will do it)
Or does "normalizing a normalized object" differ from doing it once ?

seanlip commented on core/domain/exp_domain_test.py:
```diff
@@ -717,14 +717,16 @@ def test_validation(self):
-                }
-        })
+        solution = state_domain.Solution(
+            interaction_id=init_state.interaction.id,
```
You shouldn't need to specify the argnames. The only exception to this is when a default arg is specified in the definition of the function being called.

Please fix, here and elsewhere (I saw this in multiple files in the PR).

seanlip replied:
Hi @FlorinBalint -- one note, please don't resolve the github comments (it just makes them harder for reviewers to read). Instead, say "Done" or similar if you've completed them. Thanks!

seanlip reviewed (changes requested):
Thanks @FlorinBalint! Just one comment, otherwise LGTM.

Re the normalization, I think you can leave it as-is. The test is testing a behaviour that should not occur anyway. If we wanted to add a check, then we should instead add an assertion that the value is of the correct type (rather than forcing it to the correct type, since that would hide an error that occurred earlier in the pipeline). But I think that doing this is optional.

FlorinBalint committed:
Omit argument names, decrease verbosity.

FlorinBalint commented:
Should we also make a safety copy of the passed argument ? 
(if so, should be a deepcopy? we basically had a shallow copy before)
We previously did not keep any reference to the passed dictionary.

seanlip commented:
Hm, probably not. Although it would be more correct to do that, strictly speaking, for practical reasons I don't think we need to (otherwise you end up with a copy.deepcopy() overhead at each layer of the stack). If we do decide to do it, though, we should do it for all the functions.

seanlip reviewed (approved):
LGTM. Thanks @FlorinBalint!

seanlip commented:
@kevinlee12 @nithusha21 @aks681 PTAL as codeowners. Thanks!

kevinlee12 reviewed (approved):
lgtm for code owner files.

FlorinBalint committed:
Add a null check for null solution updates.
This check makes sure that the refactored code behaves exactly the same.

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271
The 'check_e2e_tests_are_captured_in_ci_test' test is fixed on latest.

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

seanlip commented:
@aks681 @nithusha21 PTAL as codeowners.

@FlorinBalint You might want to merge develop into this branch in order to resolve the merge conflicts. Thanks!

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

nithusha21 reviewed (approved):
LGTM! Thanks @FlorinBalint

DubeySandeep commented on core/domain/exp_services.py:
```diff
@@ -378,7 +378,12 @@ def apply_change_list(exploration_id, change_list):
+                    solution = None
+                    if change.new_value:
+                        solution = state_domain.Solution.from_dict(
+                            state.interaction.id, change.new_value
```
I was expecting the params to be a `dict`.

DubeySandeep replied:
I cross-checked the from_dict method for the Solution object and I confirm it takes two params.
![image](https://user-images.githubusercontent.com/16653571/74780933-94b34400-52c6-11ea-8722-5a215c0d518e.png)

[Marking this comment as resolved]

DubeySandeep commented on core/domain/exp_services.py:
```diff
@@ -378,7 +378,12 @@ def apply_change_list(exploration_id, change_list):
                         exp_domain.STATE_PROPERTY_INTERACTION_SOLUTION):
-                    state.update_interaction_solution(change.new_value)
+                    solution = None
+                    if change.new_value:
```
`if change.new_value is not None` (Pythonic) [If we just want to pass this for "not None" case]

FlorinBalint replied:
Done

DubeySandeep commented on core/domain/state_domain.py:
```diff
@@ -1790,29 +1790,20 @@ def update_interaction_hints(self, hints_list):
-            self.interaction.solution = Solution.from_dict(
-                self.interaction.id, solution_dict)
+        if solution is not None:
+            self.interaction.solution = solution
```
Do we still need this line inside the `if-else` statement?

FlorinBalint replied:
Moved the assignment outside the if-else.

DubeySandeep commented on core/domain/state_domain.py:
```diff
@@ -1790,29 +1790,20 @@ def update_interaction_hints(self, hints_list):
             old_content_id_list, new_content_id_list)
 
-    def update_interaction_solution(self, solution_dict):
+    def update_interaction_solution(self, solution):
```
Earlier we don't have to know the id of the state to update its solution, as the method used to share the state's interaction id while creating solution object. Now we have to explicitly pass the id for creating the solution object.

@seanlip, do you think it will be a good idea to allow a solution object here which can formed using any interaction_id (not state-specific)? (I can't think of any bad condition but I'm not sure whether this will be a good idea.)

@FlorinBalint, Can you please check whether validation fails if we pass the wrong interaction_id while updating the solution? (If such validation exists, can you please add a test for that (maybe in the service layer [not sure])?

seanlip replied:
I think you need the interaction ID when creating the solution, because IIRC there is a check to ensure that the "answer" part of the solution is appropriate for the specified interaction. So I don't think it's a good idea to remove the dependence on interaction ID.

However, I support the suggestion for a test to ensure that the case @DubeySandeep suggests in the last paragraph doesn't happen. (Though it might already be handled by the solution validity checker I mentioned above? Not sure.)

seanlip replied:
/cc @ankita240796 since she implemented prod-validation jobs for this and might know if the constraint is being checked there, too -- otherwise we ought to add it.

ankita240796 replied:
There is no extra method added in prod_validation for this. The prod validator uses the validate function defined in domain objects along with additional checks specific to model properties.

FlorinBalint replied:
Sorry for not being responsive these days.

Some tests definitely failed before I added the right interaction_id 
(as an intermediate failing step).
I added a test in which the interaction does not have the id set, for the changes in exp_services.

DubeySandeep reviewed (commented):
@FlorinBalint, Thanks for the PR, I've left a few questions related to the changes below, PTAL!

aks681 reviewed (approved):
Lgtm as codeowner.

seanlip commented:
@ankita240796 -- just a heads-up, I assigned you to this PR because of [this comment](https://github.com/oppia/oppia/pull/8548#discussion_r381157564). Could you PTAL?

ankita240796 commented:
Sorry, missed it earlier @seanlip. Replied now. Thanks!

kevinlee12 commented:
@seanlip can we merge this?

seanlip commented:
I think @FlorinBalint needs to address (or at least reply to) some of the existing open comments?

FlorinBalint committed:
Add a test for invalid interaction_id in a solution update.

+ Address other review comments.

seanlip commented:
@FlorinBalint, could you please merge develop into your branch? The e2e test failures seem somewhat specific and I suspect it's because this PR has been open for a while, during which time new functionality has been added.

Also @DubeySandeep PTAL at @FlorinBalint's responses. Thanks!

DubeySandeep commented on core/domain/exp_services_test.py:
```diff
@@ -4052,10 +4052,45 @@ def test_update_interaction_solutions(self):
             exploration.init_state.interaction.solution.to_dict(),
             solution)
 
+    def test_update_solution_no_interaction_id(self):
```
Thanks for adding this test, can you add one more test for `invalid interaction id` i.e, the expected interaction id was X but passed Y.

FlorinBalint replied:
Sorry for the delays.

The test that I added fails with _key_error_ inside the _exp_services_ code, while trying to create the **Solution** object with _from_dict_.

It is not testing that the "update" has the wrong interaction_id (it can't find it), but rather that it does not have it set at all. 
Maybe I could find a better name for the test (_forgot_interaction_id_update_ ??).

I tried to do what you are asking but I think that the **"expected X but passed Y"** case is not really reachable from this part of the code. 
The interaction ids are also refreshed at each _Registry.get_interaction_by_id_ call

The code always takes the interaction_id of the state (based on state name). We already test the _"wrong state name passed"_ case (it is a common step for all types of updates).

DubeySandeep replied:
> The test that I added fails with key_error inside the exp_services code,

Can you show me the stack trace? (I wanted to see the line of code where it fails?)

If it's not possible to test this through the service layer, can we add this test to the domain layer of State object, which will test the update_interaction_sultion function? (maybe)

FlorinBalint replied:
This is the stacktrace:
```
Traceback (most recent call last):
  File "/oppia/core/domain/exp_services_test.py", line 4090, in test_update_solution_no_interaction_id
    'Changed interaction_solutions.'
  File "/oppia/core/domain/exp_services.py", line 878, in update_exploration
    updated_exploration = apply_change_list(exploration_id, change_list)
  File "/oppia/core/domain/exp_services.py", line 384, in apply_change_list
    state.interaction.id, change.new_value
  File "/oppia/core/domain/state_domain.py", line 249, in from_dict
    interaction_id).normalize_answer(
  File "/oppia/core/domain/interaction_registry.py", line 86, in get_interaction_by_id
    return cls._interactions[interaction_id]
KeyError: None
```

As for the _update_interaction_solution_ method:
The method now already receives a **Solution** object.  The object must be created from outside and this _"check"_ (no key error) is done only at the time you create the object.

The _interaction_id_ parameter forces us to create objects from outside (where you have access to it). As long as the state of the change is correct, the update will go through (otherwise it will fail sooner).

correct state  == exists and it has an interaction id 
I added a test for the second condition and there are tests for the first (the update fails faster in that case).

DubeySandeep replied:
Thanks for the explanation, I think we are having some miscommunication here, I wanted to check this:

Assume a state has interaction_id TextInput (valid intractionId), let's create a Solution object using MultipleChoiceInput (valid interaction) and pass this solution to state.update_interaction_solution and check whether this update fails? Let me know if it's still not clear?

FlorinBalint replied:
Sorry for the misunderstanding.
I had to do some debugging to understand what the interaction_id actually means, the naming is quite confusing (as it is with the way they are populated, inside the registry). It's more like an interaction type (--> I thought that there are many interactions of the same type and the different **IDs** and then there's some equals check somewhere for each interaction).

I created the test case that you are talking about and it does not fail, even though the solution has now a new type. It just replaces the solution to point to our new object.

DubeySandeep replied:
Thanks for taking a deeper look @FlorinBalint and sharing your findings! :)

[Wanted to make sure that you have validate the exploration after saving the new (buggy) solution?] 

So I think we need validation to make sure that the solution is created with a give interaction is? Can we do something for that? Also, can you think of a case where this can be an issue i.e, a solution created via a different interactionID? (If there isn't an issue then we can leave it as - is. @seanlip, do you have any suggestions?)

DubeySandeep replied:
@FlorinBalint, I had a discussion with sean for the same and I think we should add validation in the state object to make sure "if normalize(solution.interaction.id, solution.answer) != solution.correct_answer", raise an error. Can you please whether we can implement this?

Let me know if you have any questions? Thanks! :)

DubeySandeep reviewed (approved):
LGTM! Just one minor comment (below). Feel free to merge this PR once that comment is resolved.

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

kevinlee12 commented:
@DubeySandeep ptal! 

@FlorinBalint please @-mention reviewers when you need their attention.

kevinlee12 commented:
@FlorinBalint please merge from develop!

FlorinBalint committed:
Merge branch 'develop' into interaction_update_6271

Commit:
Fix #8808: Disables collapsible block and tabs RTEs for questions (#8810)

* changes for content

* Add context check for content/interaction outcome feedback/hints/solution editors to show/hide complex RTEs

* == -> ===

* address review comments

FlorinBalint committed:
Fix the new added test (from upstream) to use Solution objects instead of dict.

DubeySandeep committed:
Replace codeowners for a few files (#8867)

rohitkatlaa committed:
Completes the pending changes for #8639 (#8873)

* adding requested changes

* remove lint disable

aks681 committed:
Added the skillreview RTE element to explorations in stories (#8617)

* added model and service functions

* modified exp editor to allow skill cards

* fixed player issue

* lint

* minor change

* fixed tests

* fixed coverage

* made review changes and modified skillreview RTE

* additional review changes

* added coverage

* addressed review changes

* made review changes

* made review changes

* minor fix

* addressed comment

* Added e2e test

* made review changes

* made review changes

* fixed frontend test

Commit:
Fix #8815: Initialise Guppy symbols during exploration init (#8850)

* Initialise Guppy symbols during exploration init

* separate logic to a method

DubeySandeep committed:
Fix #8817: Fixes progress bar issue in firefox (#8881)

* Fixes progress bar issue in firefox.

* Fix minor issue.

kevinlee12 commented:
@DubeySandeep ptal!

aks681 committed:
fix #8819 (#8894)

nithusha21 committed:
Transfer codeowner for e2e test files to Chris (#8896)

* Transfer codeowner for e2e test files to Chris

* fix typo

U8NWXD committed:
Await Visible Confirmation Message on Role Update (#8889)

When an E2E test updates a role (e.g. in the
topicAndStoryEditorFileUploadFeatures suite), we do not wait for the
confirmation message to be visible. This can lead the test to proceed
even though the role update has not completed, yielding a 401 error that
breaks the test. This commit waits for the confirmation to be visible to
resolve this issue.

nishantwrp committed:
Fix #8877: Remove DebugInfoTrackerService from the codebase  (#8879)

* Remove DebugInfoTrackerService from the codebase

* Fix coverage tests

DubeySandeep reviewed (changes requested):
[Marking this "Request changes" to avoid accidental merge.]

Hudda committed:
Fixes #8625: Split linters to run for a specific file extension (#8656)

* Splited linter by file extension types

* Run linter using semaphore method

* Used Semaphore method for linting

* Fix yarn.lock

* Added more isolation

* Fix yarn.lock

* Added manager for third party lint checks

* Fix typo

* Splitted the linter into sub linters

* Moved under linter directory

* Addressed review changes

* Changed pre_commit_linter path

* Temporarily disabled BlankLineBelowFileOverviewChecker

* Fixed lint error

* Added test for redirect_stdout

* Removed html directive check

* Addressed review comments

* Used list.extend

* Fixed lint errors

* Addressed review comments

* Increased concurrency limit to 2

* Addressed review comments

* Fixed lint errors

* Semaphore fixed

* Addressed review comments

* Addressed review comments.

* Addressed review comments

* Addressed review comments

* Addressed review comments

U8NWXD committed:
Fix Continue Button Display Style (#8878)

The `continue-button` element inherited a value of `inline` for the
`display` CSS property, which caused an incorrect outline when in focus.
This commit fixes this issue by setting `display: block`.

Fixes #8826.

marianadasilvadev committed:
Add missing review comments from #8775 (#8857)

* Separate controller from $uibModal and test both controller and service to get to 100% coverage.

* Add fully covered tests in whitelist

* Add review fix

Commit:
Transfer codeowner for some LaCE team files to Kevin (#8906)

mzaman07 committed:
Fix #8802: Google Chrome on Chromebook not displaying the Speech Synthesised Voiceover. (#8892)

* Changed the speech synthesis default language to be en-US for desktop as it's widely available. This is the short-term fix.

* Updated the test to match the widely accepted speech synthesis language.

* Forgot to change these.

DubeySandeep committed:
Reviewer assignment functionality for community contribution (#8862)

sseth99 committed:
Fix #7988: Align the Question content in the table (#8917)

* {{Add classroom option to topnavbar and hide it under SHOW_CLASSROOM_CALLOUT}}

* :Revert "Merge branch 'develop' of https://github.com/oppia/oppia into develop"

This reverts commit 360c0e348081664d06a6b9c1812c892a7a64ed60, reversing
changes made to f10f1264c74c90f0091381ccf56b0eeff6f8e237.

* Add classroom option to top navbar

* Revert "Add classroom option to top navbar"

This reverts commit 20bee9294feb719014f655b5d5477087777b6399.

* Changes to add classroom option to top navbar

* Changes to add classroom option to top navbar

* Changes to add classroom option to top navbar

* Changes to add classroom option to top navbar

* Test Commit

* Commit

* Change alignment of question-content

aks681 committed:
Fix #8827, #8904 and #8903: Fixed highlight reset of feedback icon, visibility of dest state name in outcome editor and behavior of dropdown in 'Add Answer Group' modal (#8918)

* Fix #8827

* Fix #8904

* Fix #8903

Commit:
Fix #8882: Fix collections suite flakiness (#8900)

* Fix collections suite flakiness

* Address review comments

* removed waitFor.pageToFullyLoad()

* Address review comments

* Add TODOs

abhayraghuwanshi committed:
Fixes part of #8472 . Migrate AngularJS files to Angular 8 for SkillRightBackendApiService file (#8721)

* all saved

* change into angualr

* added

* service moved

* service migrated

* merge conflict solved

* error

* yarn

* file added

* added space after colon and change the order of tests

* moved the deepclone below angular import

Co-authored-by: Abhay kumar raghuwanshi <abhayraghuwanshi@users.noreply.github.com>

ankita240796 committed:
Fix path in restore backup script  (#8913)

jimbyo committed:
Fix part of #4057: Added test cases for question-backend-api.service.ts (#8916)

* added test cases for question backend api service

* added test cases for question backend api service

* fixing lint errors for question-backend-api

* fixing lint errors for question-backend-api

* add question-backend-api to FULLY_COVERED_FILENAMES

marianadasilvadev committed:
Increase coverage for improvement-modal.service.ts (#8799)

* Separate controllers from $uibModal.open in new files and test both controllers and service to get 100% coverage.

* Add fully covered files in whitelist.

* Fix review comments

* Fix review comments

* Fix review comments

* Fix review comments

* Add threadId as parameter

* Fix review comments

* Refactoring controllers

* Fix controllers in order to run e2e tests

nithusha21 committed:
fix padding for the exploration feedback (#8898)

nishantwrp committed:
Import wave-surfer using webpack instead of including it in third_party.js file (#8922)

* Import wave-surfer using webpack instead of including it in third_party.js file

* Codeowners

* Removed Relative Imports

* Add file to coverage tests

* Load WaveSurfer in only one file

* var -> const

nishantwrp committed:
Remove angular-drag-and-drop-lists.min.js from header js imports (#8929)

bansalnitish committed:
Updates to provide only required services in App.ts (#8807)

* Removes UpgradedServices from App.ts

* Fixes frontend tests

* Remove unnecessary use of UpgradedService

* Fixes e2e tests

* Fixed travis

* Fixes frontend tests

* Fixes travis tests

FlorinBalint committed:
Add a validation for the solution answer of the update method.

nishantwrp reviewed (changes requested):
@FlorinBalint I think you did not push correctly. I'd suggest closing this PR and making a new one.

FlorinBalint commented:
Interactive rebase failure: I rebased instead of merging.

copied / continued  here:
https://github.com/oppia/oppia/pull/8932

seanlip commented:
Thanks @FlorinBalint. Just a note, please use only merging from now on, as described [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change). We've had issues with rebasing in the past (it can rewrite history and cause older review comments to get lost).
